src/batch/src/executor/join/lookup_join.rs

```diff
@@ -0,0 +1,592 @@
+            .source_templates
+            .iter()
+            .map(|source| ProstExchangeSource {
+                task_output_id: Some(source.get_task_output_id().unwrap().clone()),
+                host: Some(source.get_host().unwrap().clone()),
+                local_execute_plan: Some(Plan(local_execute_plan.clone())),
+            })
+            .collect();
+
+        Ok(Box::new(DefaultGatherExecutor { prost_sources }))
+    }
+
+    /// Execute the `GatherExecutor` and retrieve the data chunk stream
+    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
+        self.get_probe_side_chunk()
+    }
+}
+
+/// Lookup Join Executor.
+/// High-level idea:
+/// 1) Iterate through each row in the build side
+/// 2) For each row R, find the rows that match with R on the join conditions on the probe side by
+/// using the `GatherExecutor` to query the probe side
+/// 3) Join R and the matching rows on the probe side together with either inner or left outer join
+/// 4) Repeat 2-3) for every row in the build side
+pub struct LookupJoinExecutor<GE> {
+    join_type: JoinType,
+    condition: Option<BoxedExpression>,
+    build_side_source: RowLevelIter,
+    build_side_data_types: Vec<DataType>,
+    probe_side_data_types: Vec<DataType>,
+    probe_side_source_params: ProbeSideSourceParams,
+    chunk_builder: DataChunkBuilder,
+    schema: Schema,
+    output_indices: Vec<usize>,
+    last_chunk: Option<SlicedDataChunk>,
+    identity: String,
+    unused: PhantomData<GE>, // Only exists so we can use generic type
+}
+
+impl<GE: 'static + GatherExecutor> Executor for LookupJoinExecutor<GE> {
+    fn schema(&self) -> &Schema {
+        &self.schema
+    }
+
+    fn identity(&self) -> &str {
+        &self.identity
+    }
+
+    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
+        self.do_execute()
+    }
+}
+
+impl<GE: 'static + GatherExecutor> LookupJoinExecutor<GE> {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(mut self: Box<Self>) {
+        self.build_side_source.load_data().await?;
+
+        loop {
+            let cur_row = self.build_side_source.get_current_row_ref();
+            if cur_row.is_none() {
+                break;
+            }
+            let cur_row = cur_row.unwrap();
+
+            let gather_executor = GE::new_boxed(&cur_row, &self.probe_side_source_params)?;
+            let mut probe_side_stream = gather_executor.execute();
+
+            let mut chunk_added = false;
+
+            loop {
+                let probe_side_chunk = probe_side_stream.next().await;
+                let chunk_is_none = probe_side_chunk.is_none();
+
+                // If None is received for the first time on the probe_side_stream AND it's a Left
+                // Outer Join AND no chunks have been added for the current row, don't break yet
+                // as we need to add NULL-padded values
+                if chunk_is_none && (self.join_type == JoinType::Inner || chunk_added) {
```

Comment:
I'm a little confusing here, why we can break after `chunk_added`?

Reply:
We can break after a `chunk_is_none` and `chunk_added` because since a chunk has already been added, we don't need to add any NULL-padded rows in the case of Left Outer Join.